lib/vector-common/src/estimated_json_encoded_size_of.rs

```diff
@@ -0,0 +1,1039 @@
+    where
+        S: ser::Serializer,
+    {
+        match &self.0 {
+            // The timestamp is converted to a static epoch timestamp, to avoid any unnecessary
+            // allocations.
+            //
+            // The following invariants must hold for the size of timestamps to remain correct:
+            //
+            // - `chrono::SecondsFormat::AutoSi` is used to calculate nanoseconds precision.
+            // - `chrono::offset::Utc` is used as the timezone.
+            // - `use_z` is `true` for the `chrono::DateTime#to_rfc3339_opts` function call.
+            Value::Timestamp(ts) => {
+                let ns = ts.nanosecond() % 1_000_000_000;
+                let epoch = if ns == 0 {
+                    EPOCH_RFC3339_0
+                } else if ns % 1_000_000 == 0 {
+                    EPOCH_RFC3339_3
+                } else if ns % 1_000 == 0 {
+                    EPOCH_RFC3339_6
+                } else {
+                    EPOCH_RFC3339_9
+                };
+
+                serializer.serialize_str(epoch)
+            }
+
+            // Collection types have their inner `Value`'s wrapped in `JsonEncodedValue`.
+            Value::Object(m) => serializer.collect_map(m.iter().map(|(k, v)| (k, Self(v)))),
+            Value::Array(a) => serializer.collect_seq(a.iter().map(Self)),
+
+            // The `Value` type serializes `Value::Bytes` using `serialize_str`, but this has two
+            // downsides:
+            //
+            // 1. For invalid UTF-8 encoded bytes, it will replace them with `U+FFFD REPLACEMENT,
+            //    requiring allocations before counting the bytes.
+            //
+            // 2. Even for valid UTF-8 encoded bytes, it will have to validate all individual bytes,
+            //    which our soaks have shown to cause a significant drop in throughput.
+            //
+            // Because of this, we take the assumption that all bytes passed through this serializer
+            // are valid UTF-8 encoded bytes, and thus can be counted as-is. If this is not the
+            // case, the final byte size will be off slightly, or significantly, depending on how
+            // many of the bytes need to be escaped, or replaced.
+            Value::Bytes(b) => serializer.serialize_bytes(b),
+
+            // All other `Value` variants are serialized according to the default serialization
+            // implementation of that type.
+            v => v.serialize(serializer),
+        }
+    }
+}
+
+/// A helper trait that is implemented for any `T` that implements `serde::Serialize`, to get the
+/// estimated JSON encoded size of that type.
+pub trait EstimatedJsonEncodedSizeOf {
+    fn estimated_json_encoded_size_of(&self) -> usize;
+}
+
+impl<T> EstimatedJsonEncodedSizeOf for T
+where
+    T: serde::Serialize,
+{
+    /// Returns the estimated JSON encoded size of `self`, or `0` if the size cannot be calculated
```

Comment:
It might be worth returning a `Result` to make the error case explicit. It's not clear to me that 0 is a good default in all cases. Just a thought, I haven't read through all the code yet.

Reply:
That's a fair point. I'll make some changes here.